src/environs/__init__.py

```diff
@@ -378,16 +377,34 @@ class TimeDeltaField(ma.fields.TimeDelta):
     def _deserialize(self, value, *args, **kwargs) -> timedelta:
         if isinstance(value, timedelta):
             return value
-        match = _TIMEDELTA_PATTERN.match(value)
-        if match is not None and match.group(0):  # disallow "", allow "0s"
+        if isinstance(value, str):
+            if value.strip() == "":
+                raise ma.ValidationError(
+                    "An empty string is not a valid period of time."
+                )
+            match = _TIMEDELTA_PATTERN.match(value.strip())
+            if match is not None:
+                return timedelta(
+                    weeks=int(match.group(1) or 0),
+                    days=int(match.group(2) or 0),
+                    hours=int(match.group(3) or 0),
+                    minutes=int(match.group(4) or 0),
+                    seconds=int(match.group(5) or 0),
+                    milliseconds=int(match.group(6) or 0),
+                    microseconds=int(match.group(7) or 0),
+                )
+            try:
+                value = float(value)
+            except ValueError:
+                raise ma.ValidationError("Not a valid period of time.") from None
+        if isinstance(value, bool):
+            raise ma.ValidationError("Not a valid period of time.")
+        if isinstance(value, (int, float)):
+            seconds = int(value)
+            milliseconds = int(value * 1000 % 1000)
             return timedelta(
-                weeks=int(match.group(1) or 0),
-                days=int(match.group(2) or 0),
-                hours=int(match.group(3) or 0),
-                minutes=int(match.group(4) or 0),
-                seconds=int(match.group(5) or 0),
-                milliseconds=int(match.group(6) or 0),
-                microseconds=int(match.group(7) or 0),
+                seconds=seconds,
+                milliseconds=milliseconds,
             )
         return super()._deserialize(value, *args, **kwargs)
```

Comment:
looks like a lot of logic was added just to bypass the parent class' validation (this line won't be hit anymore).

- a float as default value was previously slipping through the cracks (silent downcast to int)
- a float-in-string as default value didn't slip through the cracks (expected behaviour to reject)

is this PR something that should rather fold into the bigger topic of #270 and/or #297 instead?

Reply:
opened a minimal version to avoid breaking change: https://github.com/sloria/environs/pull/369

Reply:
> (this line won't be hit anymore).

It will if you provide something that is of other type than str, float, int, bool, timedelta.
So this is more a last option catch-all.

We could definitely do the more minimal change to get to the previously working (but unconsistent/unlogical behavior)... but I'm not sure what's wrong with this PR which supports the previously working pattern and actually fixes some of the inconsistent behaviors?

Reply:
and actually, I think that #297 is not relevant anymore as it was already fixed (prior to this PR)